File: db-async-common/src/main/scala/com/github/mauricio/async/db/pool/TimeoutScheduler.scala
package com.github.mauricio.async.db.pool

import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.{TimeUnit, TimeoutException, ScheduledFuture}
import io.netty.channel.EventLoopGroup
import scala.concurrent.{ExecutionContext, Promise}
import scala.concurrent.duration.Duration

trait TimeoutScheduler {

  private var isTimeoutedBool = new AtomicBoolean(false)

  /**
   *
   * The event loop group to be used for scheduling.
   *
   * @return
   */

  def eventLoopGroup : EventLoopGroup

  /**
   * Implementors should decide here what they want to do when a timeout occur
   */

  def onTimeout    // implementors should decide here what they want to do when a timeout occur

  /**
   *
   * We need this property as isClosed takes time to complete and
   * we don't want the connection to be used again.
   *
   * @return
   */

  def isTimeouted : Boolean =
    isTimeoutedBool.get

  def addTimeout[A](
                     promise: Promise[A],
                     durationOption: Option[Duration])
                   (implicit executionContext : ExecutionContext) : Option[ScheduledFuture[_]] = {
    durationOption.map {
      duration =>
        val scheduledFuture = schedule(
        {
          if (promise.tryFailure(new TimeoutException(s"Operation is timeouted after it took too long to return (${duration})"))) {
            isTimeoutedBool.set(true)
            onTimeout
          }
        },
        duration)
        promise.future.onComplete(x => scheduledFuture.cancel(false))

        scheduledFuture
    }
  }

  def schedule(block: => Unit, duration: Duration) : ScheduledFuture[_] =
    eventLoopGroup.schedule(new Runnable {
      override def run(): Unit = block
    }, duration.toMillis, TimeUnit.MILLISECONDS)
}

File: db-async-common/src/main/java/com/github/jasync/sql/db/pool/TimeoutScheduler.kt
package com.github.jasync.sql.db.pool

import com.github.jasync.sql.db.util.nullableMap
import com.github.jasync.sql.db.util.onCompleteAsync
import com.github.jasync.sql.db.util.tryFailure
import io.github.oshai.kotlinlogging.KotlinLogging
import io.netty.channel.EventLoopGroup
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicBoolean

private val logger = KotlinLogging.logger {}

interface TimeoutScheduler {

    /**
     *
     * We need this method because isClosed takes time to complete and
     * we don't want the connection to be used again.
     *
     * @return true if connection has a query that time-out
     */

    fun isTimeout(): Boolean
}

class TimeoutSchedulerImpl(
    private val executor: Executor,
    private val eventLoopGroup: EventLoopGroup,
    private val timeoutFun: () -> Unit

) : TimeoutScheduler {

    /**
     * Implementors should decide here what they want to do when a timeout occur
     */
    private fun onTimeout() {
        timeoutFun()
    }

    private var isTimeoutBool = AtomicBoolean(false)

    override fun isTimeout(): Boolean = isTimeoutBool.get()

    private fun Duration.schedule(block: () -> Unit): ScheduledFuture<*> {
        return eventLoopGroup.schedule({
            block()
        }, this.toMillis(), TimeUnit.MILLISECONDS)
    }

    fun <A> addTimeout(
        promise: CompletableFuture<A>,
        durationOption: Duration?,
        connectionId: String
    ): ScheduledFuture<*>? {
        return durationOption.nullableMap { duration ->
            logger.trace { "adding timeout $duration for connectionId $connectionId" }
            val scheduledFuture = duration.schedule {
                if (promise.tryFailure(TimeoutException("Operation timeout after it took too long to return ($duration)"))) {
                    logger.trace { "operation timeout for connectionId $connectionId" }
                    isTimeoutBool.set(true)
                    onTimeout()
                }
            }

            promise.onCompleteAsync(executor) { scheduledFuture.cancel(false) }
            scheduledFuture
        }
    }
}
